db/badger.go

```diff
@@ -68,41 +68,38 @@ func (t *badgerTxn) Delete(key []byte) error {
 }
 
 // Get : see db.Transaction.Get
-func (t *badgerTxn) Get(key []byte) ([]byte, error) {
+func (t *badgerTxn) Get(key []byte, cb func([]byte) error) error {
 	item, err := t.badger.Get(key)
 	if err != nil {
 		if errors.Is(err, badger.ErrKeyNotFound) {
-			return nil, ErrKeyNotFound
+			return ErrKeyNotFound
 		}
 
-		return nil, err
+		return err
 	}
 
-	var value []byte
-	return value, item.Value(func(val []byte) error {
-		value = append([]byte{}, val...)
-		return nil
+	return item.Value(func(val []byte) error {
+		return cb(val)
```

Comment:
Maybe I am missing something but in the docs of badger they say you must copy. From [docs (Using key/value pairs)](https://dgraph.io/docs/badger/get-started/#using-key-value-pairs) 

```go
err := db.View(func(txn *badger.Txn) error {
  item, err := txn.Get([]byte("answer"))
  handle(err)

  var valNot, valCopy []byte
  err := item.Value(func(val []byte) error {
    // This func with val would only be called if item.Value encounters no error.

    // Accessing val here is valid.
    fmt.Printf("The answer is: %s\n", val)

    // Copying or parsing val is valid.
    valCopy = append([]byte{}, val...)

    // Assigning val slice to another variable is NOT OK.
    valNot = val // Do not do this.
    return nil
  })
  handle(err)

  // DO NOT access val here. It is the most common cause of bugs.
  fmt.Printf("NEVER do this. %s\n", valNot)

  // You must copy it to use it outside item.Value(...).
  fmt.Printf("The answer is: %s\n", valCopy)

  // Alternatively, you could also use item.ValueCopy().
  valCopy, err = item.ValueCopy(nil)
  handle(err)
  fmt.Printf("The answer is: %s\n", valCopy)

  return nil
})

```

Reply:
```
  // You must copy it to use it outside item.Value(...).
```

we are not using it outside the scope of item.Value

Reply:
Also, since we are going to use pebble. It might be a good idea to drop this commit?

We should be wary of having callbacks calling callbacks. 

And since pebble has reduced our memory usage so much maybe the low-hanging fruit doesn't need to be addressed right now.

Reply:
```
    // Copying or parsing val is valid.
```

we are simply delegating parsing work to `cb`

Reply:
> We should be wary of having callbacks calling callbacks.

there is nothing wrong with that. We should keep it, there is no down side.

> much maybe the low-hanging fruit doesn't need to be addressed right now.

it is already done tho

Reply:
> there is nothing wrong with that. We should keep it, there is no down side.

[callback hell](http://callbackhell.com/) is a thing which is why javascript introduced promises. Anyway, for us it is fine.

> we are simply delegating parsing work to `cb`

In our case, we are using `cb` mostly for unmarshalling. Are we sure that unmarshall functions don't mutate `val`?

Reply:
> > there is nothing wrong with that. We should keep it, there is no down side.
> 
> [callback hell](http://callbackhell.com/) is a thing which is why javascript introduced promises. Anyway, for us it is fine.

Yeah, I am not fond of callbacks either, but in this case it has performance implications to not use them. We should avoid GC as much as we can.

 
> > we are simply delegating parsing work to `cb`
> 
> In our case, we are using `cb` mostly for unmarshalling. Are we sure that unmarshall functions don't mutate `val`?

There is no reason they would do that. Even if they do, they are mutating a temporary copy. `val` is still in the memory, it is not directly mapped to the disk. So changing `val` doesnt really do anything.